File: docusign-sdk-sample-objc/docusign-sdk-sample-objc/Managers/TemplatesManager.m
//
//  TemplateManager.m
//  docusign-sdk-sample-objc
//
//  Copyright © 2017 DocuSign. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "TemplatesManager.h"


@interface TemplatesManager ()

@property (nonatomic, retain) DSMTemplatesManager *mDSMTemplatesManager;
@property (nonatomic, retain) NSArray<DSMEnvelopeTemplateDefinition *> *mTemplateDefinitions;

@end


@implementation TemplatesManager


#pragma mark - Static Methods

+(id)sharedInstance
{
    static TemplatesManager * sharedManager = nil;

    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedManager = [[self alloc] init];

        if (sharedManager.mDSMTemplatesManager == nil)
        {
            sharedManager.mDSMTemplatesManager = [[DSMTemplatesManager alloc] init];
        }
    });
    return sharedManager;
}


#pragma mark - Public Methods

- (void) getTemplateListWithCompletion:(void(^)(NSArray<DSMEnvelopeTemplateDefinition *> *))completionHandler
{
    // retrieve list of template definitions
    [self.mDSMTemplatesManager listTemplatesWithCompletion:^(NSArray<DSMEnvelopeTemplateDefinition *> *templateDefinitions, NSError *error)
     {
         if (error)
         {
             NSLog(@"An error occurred while fetching template definitions list: %@", error);
         }
         else
         {
             // store list of template definitions
             self.mTemplateDefinitions = templateDefinitions;
             completionHandler(templateDefinitions);
         }
     }];
}


- (BOOL) templateIsCachedWithId:(NSString *)templateId
{
    return [self.mDSMTemplatesManager cacheStateOfTemplateWithId:templateId] == DSMTemplateCacheStateCached;
}


- (void) cacheTemplateWithId:(NSString *)templateId withCompletion:(void(^)(NSString *))completionHandler
{
    // download template
    [self.mDSMTemplatesManager cacheTemplateWithId:templateId completion:^(NSError *error) {
        if (error)
        {
            NSLog(@"An error occurred downloading template: %@", error);
            completionHandler(error.localizedDescription);
        }
        else
        {
            completionHandler(nil);
        }
    }];
}


- (void) removeTemplateWithId:(NSString *)templateId
{
    // remove the specified template
    [self.mDSMTemplatesManager removeCachedTemplateWithId:templateId];
}


- (void) displayTemplateForSignature:(NSString *)templateId forController:(UIViewController *)controller
                            withTabs:(id)tabs withRecipients:(id)recipients withCustomFields:(DSMCustomFields *)customFields useOfflineFlow:(BOOL)useOffline withAttachmentAtPath:(NSString *)attachmentPath withCompletion:(void(^)(UIViewController *viewController, NSError *error))completionHandler
{
    // if an attachment was included, load PDF as NSData
    NSData * attachmentData = nil;
    if (attachmentPath != nil)
    {
        attachmentData = [[NSFileManager defaultManager] contentsAtPath:attachmentPath];
    }
    
    DSMEnvelopeDefaults *envelopeDefauts = [[DSMEnvelopeDefaults alloc] init];
    envelopeDefauts.recipientDefaults = recipients;
    envelopeDefauts.tabValueDefaults = tabs;
    envelopeDefauts.customFields = customFields;
    
    [self.mDSMTemplatesManager
     presentSendTemplateControllerWithTemplateWithId:templateId
     envelopeDefaults:envelopeDefauts
     pdfToInsert:attachmentData
     insertAtPosition:DSMDocumentInsertAtPositionEnd
     signingMode:(!useOffline) ? DSMSigningModeOnline :DSMSigningModeOffline
     presentingController:controller
     animated:YES
     completion:completionHandler];
}

@end

File: docusign-sdk-sample-swift/docusign-sdk-sample-swift/Managers/TemplatesManager.swift
//
//  TemplatesManager.swift
//  docusign-sdk-sample-swift
//
//
//  Copyright © 2017 DocuSign. All rights reserved.
//

import DocuSignSDK
import Foundation

class TemplatesManager
{
    // singleton instance
    static let sharedInstance = TemplatesManager()
    
    // DSM Template Manager
    var templatesManager: DSMTemplatesManager?
    
    // list of template definitions
    var templateDefinitions: [DSMEnvelopeTemplateDefinition]?
    
    
    //This prevents others from using the default '()' initializer for this class.
    private init() {
        
        if (templatesManager == nil)
        {
            templatesManager = DSMTemplatesManager()
        }
    }
    
    typealias templateDefinitionsCompletionHandler = (_ templateDefinitions: [DSMEnvelopeTemplateDefinition]?) -> Void
    
    func getTemplateListWithCompletion(completionHandler: @escaping templateDefinitionsCompletionHandler)
    {
        // retrieve list of template definitions
        templatesManager?.listTemplates() { (templates, error) in
            NSLog("Get Template Definitions")
            
            if let error = error {
                NSLog("Error: \(error.localizedDescription)")
            }
            else {
                self.templateDefinitions = templates
                completionHandler(templates)
            }
        }
    }
    
    
    // check if template with templateId is cached on device
    func templateIsCachedWithId(templateId: String) -> Bool {
        return (templatesManager?.cacheStateOfTemplate(withId: templateId) == DSMTemplateCacheState.cached)
    }
    
    
    //typealias cacheTemplateCompletionHandler = (_ errMsg: String?) -> Void
    
    func cacheTemplateWithId(templateId: String, completionHandler: @escaping (String?) -> Void) {
        
        templatesManager?.cacheTemplate(withId: templateId, completion: { (error: Error?) in
            if let error = error {
                completionHandler(error.localizedDescription)
            }
            else {
                completionHandler(nil)
            }
        })
    }
    
    
    func removeTemplateWithId(templateId: String)
    {
        // remove the specified template
        templatesManager?.removeCachedTemplate(withId: templateId)
    }
    
    
    func displayTemplateForSignature(templateId: String, controller: UIViewController, tabData: Dictionary<String, String>, recipientData: Array<DSMRecipientDefault>, customFields:DSMCustomFields?, onlineSign: Bool, attachmentUrl: URL?, completionHandler: @escaping ((UIViewController?, Error?) -> Void))
    {
        // load PDF data
        var pdfData: Data?
        if (attachmentUrl != nil)
        {
            do {
                pdfData = try Data(contentsOf: attachmentUrl!)
            }
            catch {
                NSLog("Error loading PDF data")
            }
        }
        
        let envelopeDefaults = DSMEnvelopeDefaults()
        envelopeDefaults.recipientDefaults = recipientData.count > 0 ? recipientData : nil
        envelopeDefaults.tabValueDefaults = tabData
        envelopeDefaults.customFields = customFields
        
        templatesManager?.presentSendTemplateControllerWithTemplate (
            withId: templateId,
            envelopeDefaults: envelopeDefaults,
            pdfToInsert: pdfData,
            insertAtPosition: .end,
            signingMode: onlineSign ? .online : .offline,
            presenting: controller,
            animated: true) { (view, error) in
                if let error = error {
                    NSLog("Error encountered during signing: \(error.localizedDescription)")
                }
                if view == nil {
                    // `view` is `nil` if all of the signers pending for signature are remote
                    // A) Envelope is sent to next remote signer, should receive `DSMSigningCompletedNotification` during online signing.
                    // B) Or in case of offline signing, envelope is successfully cached and now awaiting sync.
                    NSLog("Warning: Encountered `nil view` during signing.")
                } else {
                    // DocuSign SDK UI components are active if >=1 local signers are pending signature
                    NSLog("DocuSign Native iOS SDK - UI components active")
                }
        }
    }
}

